### ISSUE TITLE
Allow components to be removed from the Course size report when all the components are removed from a course, and add a row for reporting the Recycle bin size

### ISSUE THREAD

**Issue by Undreli-Studios** [open-lms-open-source/moodle-report_coursesize#1]
Good day team, it has been observed that the function `\report_coursesize_storecomponentcacherow` from 'locallib.php', does not handle the case when all components are removed from a course:

``` php
if ($er = $DB->get_record($table, array('component' => $r->component, 'courseid' => $r->courseid))) {
    if (strval($er->filesize) != $r->filesize) {
        $r->id = $er->id;
        if (!($DB->update_record($table, $r))) {
            return false;
        }
    }
} else if (!($DB->insert_record($table, $r))) {
    return false;
}
return true;
```

As you can see, there is code to `update_record` (when the collective filesize of the component changes), and there is code to `insert_record` (when the component is added to the course for the first time). Alas, there is no code to `delete_records` for when all instances of a component are removed from a course, i.e. when the collective filesize of the component is 0.

This presents a couple of downsides: removed resources are still presented in the report, which makes the total size not match the sum of the reported sizes of the components. Also there is no indication of sizes that come from the Recycle bin.

Here are the steps to replicate:

**Preconditions**

Create a new course. On a new site, the report would look like this:

<img width="1253" alt="Screenshot 2024-01-18 at 12 48 17 PM" src="https://github.com/open-lms-open-source/moodle-report_coursesize/assets/66765389/6efd593f-f3d1-4d87-9481-7ce30a80eec4">

**Case 1: Create a mod_resource and remove it**

1. As an admin, go to the course created in the preconditions.
2. Create a File activity, for example with a 185Kb file (or any size known to you).
3. Run the task `php admin/cli/scheduled_task.php --execute='\report_coursesize\task\calculate'`.
4. Go to the report, it ought to look like this now: <img width="1246" alt="Screenshot 2024-01-18 at 1 26 10 PM" src="https://github.com/open-lms-open-source/moodle-report_coursesize/assets/66765389/e72dd3a7-a3be-46ca-bb78-26b15162fd6a">
5. Now go to the course and remove the File activity created in step 2.
6. Run the task `php admin/cli/adhoc_task.php --execute='\core_course\task\course_delete_modules'` to finish removing the module.
7. Run the task `php admin/cli/scheduled_task.php --execute='\report_coursesize\task\calculate'`.
8. Go to the report, now it looks like this: <img width="1247" alt="Screenshot 2024-01-18 at 1 30 24 PM" src="https://github.com/open-lms-open-source/moodle-report_coursesize/assets/66765389/2c0c7e45-7491-4fb9-8254-0604726ba66a">

This image shows two errors:

- The 21Kb comes from the Recycle bin, but this is not shown in the report.
- The `mod_resource` entry in the report has a size of 185Kb ... but the course has no `mod_resource` at all, so this row of the table should not be shown. 

**Case 2: Emptying the Recycle bin**

1. For good measure, go on and empty the Recycle bin from the course created in the preconditions.
2. Run the task `php admin/cli/scheduled_task.php --execute='\report_coursesize\task\calculate'`.
3. Go to the report, a new bug is formed: <img width="1251" alt="Screenshot 2024-01-18 at 1 38 00 PM" src="https://github.com/open-lms-open-source/moodle-report_coursesize/assets/66765389/9af896c7-f77d-4553-a3f6-f4aa45db1225">

The problem in this image is that the course still holds the 21Kb from the Recycle bin, despite it being empty. But the category displays the correct 0 bytes of content.

**Note:** This same behavior has been observed when uploading files to assignments, which in the report appear as the components `assignsubmission_file` and `assignfeedback_editpdf`.

**Observed behavior**

For Case 1, the `mod_resource` entry appears even when there are no `mod_resource` in the course, and the Recycle bin size is shown but the report doesn't show that this size comes from the Recycle bin.

For Case 2, the correct 0 bytes size is shown at the category level, but the course has 21 Kb that were from the Recycle bin (but which don't exist no more), and the `mod_resource` is still present as in Case 1.

**Expected behavior**

For Case 1, the `mod_resource` entry disappears, and a Recycle bin entry shows the size of it.

For Case 2, the course only has a 0 bytes size displayed (since there is no content on the course).

**Comment by aolley** [comment]
Thanks for the detailed report!

I'll look to get this sorted today if I can.

**Comment by aolley** [comment]
This should be resolved by 5e1718c2214e1bd4d1a301a3833a3cfc4f113355
